### DISCUSSION TITLE
Ideas for multithreaded rendering

### DISCUSSION THREAD

**Discussion by rovinski** [The-OpenROAD-Project/OpenROAD#3359] [Ideas]
I saw the conversation in #3330, #3342, #3348, etc. and started thinking about how the rendering time might be improved aside from not drawing objects that aren't useful.

It seems that the rendering, for the most part, is embarrassingly parallel. In `LayoutViewer::drawInstanceOutlines()` for example (the most expensive layer measured in #3348), a function is applied to every instance in a vector to determine how / if it gets drawn, and then it makes a draw call.

https://github.com/The-OpenROAD-Project/OpenROAD/blob/6920ac530b6959f742e75b7f55062b1642730b14/src/gui/src/layoutViewer.cpp#L2041-L2091

This is essentially a [map reduce](https://en.wikipedia.org/wiki/MapReduce) paradigm which should be extremely parallel, except that Qt painter objects aren't thread safe. There was discussion in #3279 that the database is not "thread safe", but I think that was in the context of read/write operations and not read-only operations.

Beyond that, each layer is independently but sequentially rendered:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/6920ac530b6959f742e75b7f55062b1642730b14/src/gui/src/layoutViewer.cpp#L2555-L2599

There may or may not be ordering required on draw calls to these layers, I'm not sure.

I looked at a few [Qt guides](https://doc.qt.io/qt-5/qtcore-threads-mandelbrot-example.html) and they all recommend separating computation from rendering in separate threads for better performance (and interactivity).

I think some of these things were mentioned by @QuantamHD in #1930 but the PR itself seems to be unrelated (and abandoned).

I have some interest in looking into this, but I wonder what the best approach would be for this? Qt has [QThread](https://doc.qt.io/qt-5/qthread.html) and [QThreadPool](https://doc.qt.io/qt-5/qthreadpool.html) for this exact purpose. There's also [QtConcurrent](https://doc.qt.io/qt-5/qtconcurrent.html) which may have some uses.

In an ideal world, it seems like there should be a thread pool with `getThreadCount()-1` threads and the main thread. The main thread has a render queue which atomically pops the queue for draw calls and does nothing else until all worker threads yield. The worker threads can parallelize at the granularity of layer or instance/shape depending on what's optimal. But this could take a lot of refactoring, possibly be overkill, or possibly be difficult to tune the amount of threading correctly.

An intermediate step could be to simply parallelize some of the heavy-weight tasks that block the GUI thread, such as using `std::transform(std::execution::par)` or `std::for_each(std::execution::par)` and separating the draw calls from the filtering. But also these functions don't seem to be well supported on old versions of GCC/Clang (it requires Intel `libtbb` and may error or just use a serial version if it isn't linked, but I think it also think it doesn't work on Apple silicon). Maybe using OpenMP would work?

Any thoughts or opinions?
@gadfort @maliberty @oharboe @QuantamHD

**Comment by oharboe** [comment]
After the recent performance improvement, the biggest performance problem currently for us, is that all intermediate zoom levels are rendered https://github.com/The-OpenROAD-Project/OpenROAD/issues/3278

So if that could be fixed FIRST... :-)

Afterwards, the suggestions that you propose sound wicked!

It is true that Qt nor the database is thread safe, then on the other hand, this means that the GUI is single threaded, so I suppose that you are right that you can use as many or few threads as you like as long as Qt nor the database functions are non-the-wiser. Rendering should be especially amenable to this since it is a read only operation...

Also, rendering doesn't have to be perfect.  I have been thinking that a draft could be rendered while zooming and then when the user stops giving input, more detail can be rendered.

**Reply by rovinski** [reply]
I think there's a couple ways to enhance interactivity orthogonally to sheer render performance.

One is to use a cached copy of a render and apply zoom transformations to it. I think this could be as simple as using `QPixMap::scaled()` to generate intermediate frames while the new zoom level is rendered.

The other is to update the view while the render is still working. This would mean that the display doesn't "hang" until the new zoom level is fully rendered, but instead shows instances, metals, etc. popping in "slowly" when they are rendered.

If these two techniques are both used, you could get something similar to Google Maps where updates pop in as they are ready but the interface stays responsive the whole time.

That would be really cool, but also possibly more than I could handle right now :). I think setting up some kind of threaded infrastructure would be a good stepping stone.

Re: #3278 I wonder if the issue is that the update isn't checked during the render. As Matt mentioned I think the scroll events should get collapsed, but that might still mean that the first render only works off of the first scroll event and then the second render works on all the  collapsed events.

**Reply by oharboe** [reply]
W.r.t. performance, I would be *delighted* if #3278 is fixed. It is really the last practical obstacle w.r.t. performance, the rest is gravvy.

I always say that there are two universally loved features, get them right and you will have undying loyalty from your users: quality(no bugs) and performance. If it is fast and it works, it is going to be VERY hard to convince users to switch :-)

**Comment by maliberty** [comment]
@rovinski I've considered all these ideas but lack the time to implement them all.  Right now I'm looking at moving the rendering to a separate thread and making it interruptible.  The main challenge is figuring out how to trigger the right events at the right time to make Qt happy.  It is very particular that you can only paint in paintEvent.

Assuming I can land this then it is more straightforward to split that worker thread into multiple threads.  KLayout works in a similar fashion where they render per layer in each thread and composite in the main thread.

**Reply by oharboe** [reply]
This sounds like a very general general solution. Will this fix all cases, even the responsiveness of the heatmap dialog box?

**Reply by maliberty** [reply]
Ideally that's why I'm trying it.  Qt is fighting me though...

**Reply by oharboe** [reply]
But it will be worth the struggle and wait, I think. It is an elegant and generic solution that avoids cruft elsewhere, even if it requires wrestling with Qt and possibly some tricky code.

**Reply by rovinski** [reply]
I think there are opportunities within a layer as well. There probably needs to be some kind of work stealing to ensure that there isn't one thread bottlenecking the render.

If there's something in particular you want me to look into, I'd be happy to.

**Reply by Iyury1** [reply]
I also wouldnt mind helping out to implement some of these other ideas now that the renderThread is up and working.

**Reply by maliberty** [reply]
That would be great.  I would start by having each thread render one layer in its own buffer and then composite them in the main thread in order.  You'll have to keep the interrupt / restart working across the whole thread group.
